app/scripts/modules/core/src/cluster/cluster.service.ts

```diff
@@ -84,45 +84,43 @@ export class ClusterService {
     }
   }
 
+  private generateServerGroupLookupKey(serverGroup: IServerGroup): string {
+    const { name, account, region, category } = serverGroup;
+    return [name, account, region, category].join('-');
+  }
+
   public addServerGroupsToApplication(application: Application, serverGroups: IServerGroup[] = []): IServerGroup[] {
+    // map of incoming data
+    const remoteMap = keyBy(serverGroups, this.generateServerGroupLookupKey);
+    // map local cache
+    const localMap = keyBy(application.serverGroups.data, this.generateServerGroupLookupKey);
+
     if (application.serverGroups.data) {
       const data = application.serverGroups.data;
       // remove any that have dropped off, update any that have changed
-      const toRemove: number[] = [];
       data.forEach((serverGroup: IServerGroup, idx: number) => {
-        const matches = serverGroups.filter(
-          test =>
-            test.name === serverGroup.name &&
-            test.account === serverGroup.account &&
-            test.region === serverGroup.region &&
-            test.category === serverGroup.category,
-        );
-        if (!matches.length) {
-          toRemove.push(idx);
-        } else {
-          if (serverGroup.stringVal && matches[0].stringVal && serverGroup.stringVal !== matches[0].stringVal) {
-            data[idx] = matches[0];
+        const match = remoteMap[this.generateServerGroupLookupKey(serverGroup)];
+        if (match) {
+          // Match found between local and incoming data, update but only if needed
+          if (serverGroup.stringVal && match.stringVal && serverGroup.stringVal !== match.stringVal) {
+            data[idx] = match;
           }
+        } else {
+          // Not found means server group was removed: nullify in place for a single pass removal
+          data[idx] = null;
         }
       });
 
-      toRemove.forEach(idx => data.splice(idx, 1));
-
       // add any new ones
       serverGroups.forEach(serverGroup => {
-        if (
-          !application.serverGroups.data.filter(
-            (test: IServerGroup) =>
-              test.name === serverGroup.name &&
-              test.account === serverGroup.account &&
-              test.region === serverGroup.region &&
-              test.category === serverGroup.category,
-          ).length
-        ) {
+        const match = localMap[this.generateServerGroupLookupKey(serverGroup)];
+        if (!match) {
           data.push(serverGroup);
         }
       });
-      return data;
+      // filter out the removed server groups (nullified above)
+      application.serverGroups.data = data.filter(Boolean);
```

Comment:
doesn't this line undo the "update in place"?  Maybe I don't understand what references we need to retain.  In the previous version, the `data` array was `splice`'d so the array reference was the same.

Reply:
Ah, you're right. Nice catch. Let me bring back the splice.

Since removals should be far less common than matches, I don't think it will be too costly.